### PR TITLE
Optimize cart rule performances

### PR DESCRIPTION
**Pull request by jocel1**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Greaty optimize cart rules performances by making them easily cacheable
| Type?         | improvement
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? |
| How to test?  | run tests

#### Important guidelines

* Make sure [your local branch is up to date](https://help.github.com/articles/syncing-a-fork/) before commiting your changes!
* Your code MUST respect [our Coding Standards](http://doc.prestashop.com/display/PS16/Coding+Standards) (for code written in PHP, JavaScript, HTML/CSS/Smarty/Twig, SQL)!
* Your commit name MUST respect our [naming convention](http://doc.prestashop.com/display/PS16/How+to+write+a+commit+message)!

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/8402)

### CODE REVIEW

**Comment by jocel1** [comment]
Created on 1.7.2.x as it has a big impact on performances (tested with a real customer). For this customer, the average generation time drops from 408ms to 275ms (33% faster on average) according to loadimpact.com with 11 differents URLs tested (home, categories, product pages, ...) and 200 unique visitors as the same time.

File: classes/CartRule.php [hunk 1]
```diff
@@ -319,15 +347,17 @@ public static function getCustomerCartRules(
...
             return array();
         }
 
         $sql_part1 = '* FROM `' . _DB_PREFIX_ . 'cart_rule` cr
 				LEFT JOIN `' . _DB_PREFIX_ . 'cart_rule_lang` crl ON (cr.`id_cart_rule` = crl.`id_cart_rule` AND crl.`id_lang` = ' . (int) $id_lang . ')';
 
-        $sql_part2 = ' AND cr.date_from < "' . date('Y-m-d H:i:s') . '"
-				AND cr.date_to > "' . date('Y-m-d H:i:s') . '"
+        $sql_part2 = ' AND cr.date_from < "' . date('Y-m-d H:i:00') . '"
+				AND cr.date_to > "' . date('Y-m-d H:i:00') . '"
```
Comment (tomlev) [inline classes/CartRule.php]: this is wrong if date_to is the end date for the voucher : you need to use :59

File: classes/CartRule.php [hunk 2]
```diff
@@ -295,6 +295,34 @@ public static function getIdByCode($code)
...
 
     /**
+     * Check if some cart rules exists today for the given customer
+     *
+     * @param int $idCustomer
+     *
+     * @return bool
+     */
+    public static function haveCartRuleToday($idCustomer)
+    {
```
Comment (tomlev) [inline classes/CartRule.php:305]: in this method you only check if cart rule starts or ends today
not if it is valid today
Is there a reason for this restriction ?

**Review by tomlev — DISMISSED** [review]
More precisions needed please

File: classes/CartRule.php (hunk 1 above)
Comment (jocel1) [reply classes/CartRule.php]: fixed

File: classes/CartRule.php (hunk 2 above)
Comment (jocel1) [reply classes/CartRule.php:305]: I wanted to use the most generic query to make it easily cacheable

**Comment by mickaelandrieu** [comment]
Hello @tomlev, can you reconsider your review? If it's ok for you, I'll merge before the freeze.

File: classes/CartRule.php [hunk 3]
```diff
@@ -295,6 +295,34 @@ public static function getIdByCode($code)
...
+        static $haveCartRuleToday = array();
```
Comment (Quetzacoalt91) [inline classes/CartRule.php:306]: @tomlev, something to say regarding that line?

Comment (jocel1) [reply classes/CartRule.php:306]: This one is not too annoying for the tests since we just have to change the idCustomer

**Review by toutantic — APPROVED** [review]

**Pull request merged**